src/Processors/QueryPlan/Optimizations/reuseStorageOrdering.cpp

```diff
@@ -0,0 +1,129 @@
+        return 0;
+    if (sorting_node->children.size() != 1)
+        return 0;
+
+    auto * possible_read_from_merge_tree_node = sorting_node->children.front();
+
+    auto * expression = typeid_cast<ExpressionStep *>(possible_read_from_merge_tree_node->step.get());
+    if (expression)
+    {
+        if (possible_read_from_merge_tree_node->children.size() != 1)
+            return 0;
+
+        possible_read_from_merge_tree_node = possible_read_from_merge_tree_node->children.front();
+    }
+
+    auto * quota_and_limits = typeid_cast<SettingQuotaAndLimitsStep *>(possible_read_from_merge_tree_node->step.get());
+    if (quota_and_limits)
+    {
+        if (possible_read_from_merge_tree_node->children.size() != 1)
+            return 0;
+
+        possible_read_from_merge_tree_node = possible_read_from_merge_tree_node->children.front();
+    }
+
+    auto * read_from_merge_tree = typeid_cast<ReadFromMergeTree *>(possible_read_from_merge_tree_node->step.get());
+    if (!read_from_merge_tree)
+    {
+        return 0;
+    }
+
+    /// Window <- Sorting <- [Expression] <- [SettingQuotaAndLimits] <- ReadFromMergeTree
+
+    auto context = read_from_merge_tree->getContext();
+    const auto & query_info = read_from_merge_tree->getQueryInfo();
+    const auto * select_query = query_info.query->as<ASTSelectQuery>();
+
+    auto analyzer = std::make_unique<ExpressionAnalyzer>(query_info.query, query_info.syntax_analyzer_result, context);
+
+    if (select_query->window())
+    {
+        ManyExpressionActions order_by_elements_actions;
+
+        for (const auto & ptr : select_query->window()->children)
+        {
+            const auto & elem = ptr->as<const ASTWindowListElement &>();
+            auto ast = elem.definition.get();
+            const auto & definition = ast->as<const ASTWindowDefinition &>();
+
+            if (definition.partition_by)
+            {
+                for (const auto & column_ast : definition.partition_by->children)
+                {
+                    auto actions_dag = std::make_shared<ActionsDAG>(analyzer->columns_after_window);
+                    analyzer->getRootActions(column_ast, false, actions_dag);
+                    order_by_elements_actions.emplace_back(
+                            std::make_shared<ExpressionActions>(actions_dag, ExpressionActionsSettings::fromContext(context, CompileExpressions::yes)));
+                }
+            }
+
+            if (definition.order_by)
+            {
+                for (const auto & column_ast : definition.order_by->children)
+                {
+                    auto actions_dag = std::make_shared<ActionsDAG>(analyzer->columns_after_window);
+                    analyzer->getRootActions(column_ast, false, actions_dag);
+                    order_by_elements_actions.emplace_back(
+                            std::make_shared<ExpressionActions>(actions_dag, ExpressionActionsSettings::fromContext(context, CompileExpressions::yes)));
```

Comment:
This cycle is a complete copy-paste of a previous one. It can be generalised.

Reply:
Done.